dbt/task/generate.py

```diff
@@ -0,0 +1,147 @@
+
+        {
+            'test_schema': {
+                'test_table': {
+                    'metadata': {
+                        'comment': None,
+                        'name': 'test_table',
+                        'type': 'BASE TABLE',
+                        'schema': 'test_schema',
+                    },
+                    'columns': [
+                        {
+                            'type': 'integer',
+                            'comment': None,
+                            'index': bigint(1),
+                            'name': 'id'
+                        }
+                    ]
+                }
+            }
+        }
+
+    Required keys in each column: table_schema, table_name, column_index
+
+    Keys prefixed with 'column_' end up in per-column data and keys prefixed
+    with 'table_' end up in table metadata. Keys without either prefix are
+    ignored.
+    """
+    structured = {}
+    for entry in columns:
+        schema_name = entry['table_schema']
+        table_name = entry['table_name']
+
+        if schema_name not in structured:
+            structured[schema_name] = {}
+        schema = structured[schema_name]
+
+        if table_name not in schema:
+            metadata = get_stripped_prefix(entry, 'table_')
+            schema[table_name] = {'metadata': metadata, 'columns': []}
+        table = schema[table_name]
+
+        column = get_stripped_prefix(entry, 'column_')
+        # the index should really never be that big so it's ok to end up
+        # serializing this to JSON (2^53 is the max safe value there)
+        column['index'] = bigint(column['index'])
+        table['columns'].append(column)
+    return structured
+
+
+# derive from BaseTask as I don't really want any result interpretation.
+class GenerateTask(BaseTask):
+    def get_all_projects(self):
+        root_project = self.project.cfg
+        all_projects = {root_project.get('name'): root_project}
+        # we only need to load the global deps. We haven't compiled, so our
+        # project['module-path'] does not exist.
+        dependency_projects = dbt.utils.dependencies_for_path(
+            self.project, GLOBAL_DBT_MODULES_PATH
+        )
+
+        for project in dependency_projects:
+            name = project.cfg.get('name', 'unknown')
+            all_projects[name] = project.cfg
+
+        if dbt.flags.STRICT_MODE:
+            dbt.contracts.project.ProjectList(**all_projects)
+
+        return all_projects
+
+    def _get_manifest(self):
+        # TODO: I'd like to do this better. We can't use
+        # utils.dependency_projects because it assumes you have compiled your
+        # project (I think?) - it assumes that you have an existing and
+        # populated project['modules-path'], but 'catalog generate' shouldn't
+        # require that. It might be better to suppress the exception in
+        # dependency_projects if that's reasonable, or make it a flag.
```

Comment:
when you say `existing and populated project['modules-path']`, does that mean that the directory exists, and that `dbt deps` has been run to pull in the project's dependencies?

imo, `dbt deps` should be run before `dbt catalog generate` -- I can't imagine this working any other way

Reply:
Yeah. As it is, this works in the sample `talk` project I have without having to run `dbt deps` - the only thing it needs the manifest for at all is an internally defined operation and internally defined macros (right?). So it shouldn't have to look at any non-global dependencies.

But maybe that could be because it's such a simple project, I don't know.

Reply:
Yeah, it's just a side effect of it being such a simple project. There are cases where there are actually models in the dependency project that end up in the manifest, and the corresponding tables in the database should end up in the catalog as well. But I think this code is fine as written. It's perfectly fine to assert that the user has to run dbt deps before one or both of these commands will work even if it's not theoretically necessary to generate the catalog right now.